core/src/main/java/io/ably/lib/realtime/RealtimeChannelBase.java

```diff
@@ -92,13 +92,21 @@ private void setState(ChannelState newState, ErrorInfo reason, boolean resumed)
     private void setState(ChannelState newState, ErrorInfo reason, boolean resumed, boolean notifyStateChange) {
         Log.v(TAG, "setState(): channel = " + name + "; setting " + newState);
         ChannelStateListener.ChannelStateChange stateChange;
-        synchronized(this) {
+        synchronized (this) {
             stateChange = new ChannelStateListener.ChannelStateChange(newState, this.state, reason, resumed);
             this.state = stateChange.current;
             this.reason = stateChange.reason;
         }
 
-        if(notifyStateChange) {
+        if (newState == ChannelState.detached || newState == ChannelState.suspended || newState == ChannelState.failed)
+            properties.channelSerial = null; //RTP5a1
+
+        if (newState == ChannelState.attached && state == ChannelState.attached) {
```

Comment:
I'm a bit confused what the difference between `newState` and `channel.state` is here? Are these expected to be the same?

Reply:
newState is received from the server and channel.state is the current state of the channel which is old in this context.

Reply:
ok so is this only doing a re-enter if it receives `ATTACHED` if its already in the `ATTACHED` state? (which seems to contradict RTP17f?)

Reply:
I think I tried to meet expectation of the case in braces `(It does not need to do so for @RTL12@ @ATTACHED@ events received on already-@ATTACHED@ channels).`
As I can see newState and state will be the same as newState is assigned to state in code line 97 so latter check can be removed (state == ChannelState.attached) or moved before state is changed.

Reply:
I have added check to better specify `RTP17f` which will reEnter() only if new state is attached and old is not attached as described in braces text.

Old state is initiated with value of current state, before it is resigned with new state.
```
if (newState == ChannelState.attached && oldState != ChannelState.attached) {
    presence.reEnter(newState);  //RTP17f
}
```

Reply:
Actually, I do not need to check states as I already check in presence.reEnter method. 
However, I needed to change case where channel state is not attached.
```
if (newState == ChannelState.attached && channel.state != ChannelState.attached) {
    internalPresence.reEnter();
}
```

Reply:
@andydunstall can you confirm my findings, please?

Reply:
@andydunstall This was my last decision on the matter https://github.com/ably/ably-java/pull/842/commits/0287b17b8e72fc88332dac8d5b7cb33b6e3ffb25

Reply:
@qsdigor I'm not sure what you have now makes any sense:
- You're only doing a re-enter if `newState is attached && channel.state != attached`. But right above, [on line 97](https://github.com/ably/ably-java/pull/842/files#diff-6b3bce3cee6e98aa7a5f584556ac3848bd9a25c3ecc1955f171648d3978c11b6R97), the method is setting `this.state = stateChange.current`. So at that point, won't the channel state always equal the newstate because you've just set it to be?
- Having logic in setState to avoid the RTL12 'attached event when already attached' case will never fire anyway, because  [`setAttached` doesn't call setState if it's already attached](https://github.com/ably/ably-java/blob/b4eaadd3bc613535f3862a78850e5ae8f7844c30/core/src/main/java/io/ably/lib/realtime/RealtimeChannelBase.java#L315-L324) (because it doesn't need to, there's no state change)

Reply:
@SimonWoolf I am not sure if that is the case but there is a way where I can make sure to compare real channel states and makes it more readable. Here is the change https://github.com/ably/ably-java/pull/842/commits/e91d7704f919fbebe2a371bea229ffbc93c796e0

Reply:
(Discussed in standup)